### ISSUE TITLE
Ruby - cannot set socks proxy for remote Firefox

### ISSUE THREAD

**Issue by MatzFan** [SeleniumHQ/selenium#7138]
## 🐛 Bug Report

As per the [docs](https://github.com/SeleniumHQ/selenium/wiki/Ruby-Bindings#remote) I tried setting a proxy for a remote Firefox running in a Docker container on my local machine. This resulted in a `Net::ReadTimeout` error. Stack trace as follows:

```
/home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/protocol.rb:217:in `rbuf_fill': Net::ReadTimeout with #<TCPSocket:(closed)> (Net::ReadTimeout)
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/protocol.rb:191:in `readuntil'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/protocol.rb:201:in `readline'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http/response.rb:40:in `read_status_line'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http/response.rb:29:in `read_new'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:1509:in `block in transport_request'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:1506:in `catch'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:1506:in `transport_request'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:1479:in `request'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/default.rb:129:in `response_for'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/default.rb:82:in `request'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/bridge.rb:102:in `create_session'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/bridge.rb:56:in `handshake'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/driver.rb:39:in `initialize'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/driver.rb:58:in `new'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/driver.rb:58:in `for'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver.rb:88:in `for'
  from /home/me/selenium.rb:7:in `<main>'
```

## To Reproduce

Run Firefox in a docker container locally and try and instantiate a Selenium driver for the remote browser with a socks proxy set, in Ruby. My code tried to set the socks proxy so I can route traffic through Tor, but it is not necessary to actually run Tor to reproduce this error.

## Expected behavior

I would expect the driver to be instantiated.

## Test script or set of commands reproducing this issue

`$ docker run -d -p 4444:4444 -v /dev/shm:/dev/shm selenium/standalone-firefox`

<-- Ruby script - "selenium.rb" -->
```ruby
#!/usr/bin/env ruby
require 'selenium-webdriver'

caps = Selenium::WebDriver::Remote::Capabilities.firefox(
  proxy: Selenium::WebDriver::Proxy.new(socks: '127.0.0.1:9050')
)
Selenium::WebDriver.for(:remote, desired_capabilities: caps)
```

## Environment

OS: Ubuntu 18.04
Browser: Firefox 
Browser version: 66.0.3
Browser Driver version: GeckoDriver 0.24.0
Language Bindings version: Ruby 3.142.0

**Comment by p0deje** [comment]
Can you also try setting `socks_version` on proxy?

**Comment by p0deje** [comment]
WebDriver specification requires both to be set:

> If proxy has an own property for "socksProxy" and does not have an own property for "socksVersion" return an error with error code invalid argument.

However, it seems like GeckoDriver has a long-standing issue with (mozilla/geckodriver#1022) with no returning proper error and message.

**Comment by MatzFan** [comment]
Morning Alex, thanks for the quick response. That gives a cast error, code below.
```
[remote server] org.openqa.selenium.Proxy(Proxy.java):94:in `<init>': java.lang.Long cannot be cast to java.lang.Integer (Selenium::WebDriver::Error::UnknownError)
  from [remote server] org.openqa.selenium.remote.ProtocolHandshake$Result(ProtocolHandshake.java):151:in `lambda$static$0'
  from [remote server] org.openqa.selenium.remote.ProtocolHandshake$Result(ProtocolHandshake.java):166:in `<init>'
  from [remote server] org.openqa.selenium.remote.W3CHandshakeResponse(W3CHandshakeResponse.java):107:in `lambda$successHandler$1'
  from [remote server] org.openqa.selenium.remote.HandshakeResponse(HandshakeResponse.java):30:in `lambda$getResponseFunction$0'
  from [remote server] org.openqa.selenium.remote.ProtocolHandshake(ProtocolHandshake.java):126:in `lambda$createSession$0'
  from [remote server] java.util.stream.ReferencePipeline$3$1(ReferencePipeline.java):193:in `accept'
  from [remote server] java.util.Spliterators$ArraySpliterator(Spliterators.java):958:in `tryAdvance'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):126:in `forEachWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):498:in `copyIntoWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):485:in `copyInto'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):471:in `wrapAndCopyInto'
  from [remote server] java.util.stream.FindOps$FindOp(FindOps.java):152:in `evaluateSequential'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):234:in `evaluate'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):464:in `findFirst'
  from [remote server] org.openqa.selenium.remote.ProtocolHandshake(ProtocolHandshake.java):128:in `createSession'
  from [remote server] org.openqa.selenium.remote.ProtocolHandshake(ProtocolHandshake.java):74:in `createSession'
  from [remote server] org.openqa.selenium.grid.session.remote.RemoteSession$Factory(RemoteSession.java):147:in `performHandshake'
  from [remote server] org.openqa.selenium.grid.session.remote.ServicedSession$Factory(ServicedSession.java):161:in `apply'
  from [remote server] org.openqa.selenium.remote.server.ActiveSessionFactory(ActiveSessionFactory.java):180:in `lambda$apply$12'
  from [remote server] java.util.stream.ReferencePipeline$3$1(ReferencePipeline.java):193:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$11$1(ReferencePipeline.java):373:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$2$1(ReferencePipeline.java):175:in `accept'
  from [remote server] java.util.Spliterators$ArraySpliterator(Spliterators.java):958:in `tryAdvance'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):126:in `forEachWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):498:in `copyIntoWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):485:in `copyInto'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):471:in `wrapAndCopyInto'
  from [remote server] java.util.stream.FindOps$FindOp(FindOps.java):152:in `evaluateSequential'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):234:in `evaluate'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):464:in `findFirst'
  from [remote server] org.openqa.selenium.remote.server.ActiveSessionFactory(ActiveSessionFactory.java):183:in `apply'
  from [remote server] org.openqa.selenium.remote.server.NewSessionPipeline(NewSessionPipeline.java):66:in `lambda$null$2'
  from [remote server] java.util.stream.ReferencePipeline$3$1(ReferencePipeline.java):193:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$2$1(ReferencePipeline.java):175:in `accept'
  from [remote server] java.util.Collections$2(Collections.java):4717:in `tryAdvance'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):126:in `forEachWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):498:in `copyIntoWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):485:in `copyInto'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):471:in `wrapAndCopyInto'
  from [remote server] java.util.stream.FindOps$FindOp(FindOps.java):152:in `evaluateSequential'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):234:in `evaluate'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):464:in `findFirst'
  from [remote server] org.openqa.selenium.remote.server.NewSessionPipeline(NewSessionPipeline.java):69:in `lambda$createNewSession$3'
  from [remote server] java.util.stream.ReferencePipeline$3$1(ReferencePipeline.java):193:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$3$1(ReferencePipeline.java):193:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$3$1(ReferencePipeline.java):193:in `accept'
  from [remote server] java.util.stream.DistinctOps$1$2(DistinctOps.java):175:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$2$1(ReferencePipeline.java):175:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$3$1(ReferencePipeline.java):193:in `accept'
  from [remote server] java.util.stream.ReferencePipeline$2$1(ReferencePipeline.java):175:in `accept'
  from [remote server] java.util.stream.Streams$StreamBuilderImpl(Streams.java):405:in `tryAdvance'
  from [remote server] java.util.stream.Streams$ConcatSpliterator(Streams.java):728:in `tryAdvance'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):126:in `forEachWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):498:in `copyIntoWithCancel'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):485:in `copyInto'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):471:in `wrapAndCopyInto'
  from [remote server] java.util.stream.FindOps$FindOp(FindOps.java):152:in `evaluateSequential'
  from [remote server] java.util.stream.AbstractPipeline(AbstractPipeline.java):234:in `evaluate'
  from [remote server] java.util.stream.ReferencePipeline(ReferencePipeline.java):464:in `findFirst'
  from [remote server] org.openqa.selenium.remote.server.NewSessionPipeline(NewSessionPipeline.java):72:in `createNewSession'
  from [remote server] org.openqa.selenium.remote.server.commandhandler.BeginSession(BeginSession.java):65:in `execute'
  from [remote server] org.openqa.selenium.remote.server.WebDriverServlet(WebDriverServlet.java):235:in `lambda$handle$0'
  from [remote server] java.util.concurrent.Executors$RunnableAdapter(Executors.java):511:in `call'
  from [remote server] java.util.concurrent.FutureTask(FutureTask.java):266:in `run'
  from [remote server] java.util.concurrent.ThreadPoolExecutor(ThreadPoolExecutor.java):1149:in `runWorker'
  from [remote server] java.util.concurrent.ThreadPoolExecutor$Worker(ThreadPoolExecutor.java):624:in `run'
  from [remote server] java.lang.Thread(Thread.java):748:in `run'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/response.rb:72:in `assert_ok'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/response.rb:34:in `initialize'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/common.rb:88:in `new'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/common.rb:88:in `create_response'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/default.rb:114:in `request'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/bridge.rb:102:in `create_session'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/bridge.rb:56:in `handshake'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/driver.rb:39:in `initialize'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/driver.rb:58:in `new'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/driver.rb:58:in `for'
  from /home/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver.rb:88:in `for'
  from /home/me/Downloads/selenium.rb:10:in `<main>'
```
```ruby
#!/usr/bin/env ruby
require 'selenium-webdriver'

caps = Selenium::WebDriver::Remote::Capabilities.firefox(
  proxy: Selenium::WebDriver::Proxy.new(
    socks: '127.0.0.1:9050',
    socks_version: 5
  )
)
Selenium::WebDriver.for(:remote, desired_capabilities: caps)
```
FWIW I can instantiate a driver with an http proxy, simply by replacing `:socks` with `:http` in my original code, so the problem seems to be specific to socks.

I see the unit tests for the Proxy include `:socks_version`, but the problem is during driver instantiation again (error is at last line in my code). Might it be worth adding various proxies to the driver integration tests?

**Comment by p0deje** [comment]
The issue you see now was just fixed a couple of weeks ago in 9b52da06803f90bba9cfac89878c1366d21af33e. However, it's only released in `4.0.0-alpha-1`, so I'm not sure what's the best way for you to get the release. @barancev Could you advise?

**Comment by p0deje** [comment]
So it seems like this will only be properly fixed in 4.0.
